File: web/tests/log-excerpt.test.js
/**
 * Headless tests for web/js/lib/log-excerpt.js (WP 4a.5).
 * Run: node --test "web/tests/*.test.js"   (see web/tests/README.md)
 */
import { test } from "node:test";
import assert from "node:assert/strict";
import { selectExcerptDisplay, EXCERPT_STATE } from "../js/lib/log-excerpt.js";

test("collapsed: a row that isn't expanded shows nothing, regardless of any other field", () => {
  assert.deepEqual(selectExcerptDisplay({ expanded: false }), {
    state: EXCERPT_STATE.COLLAPSED,
    lines: [],
    truncated: false,
  });
  // Even a fetched excerpt or an in-flight load must not leak through while
  // collapsed — a row can be re-collapsed while its fetch is still in
  // flight (fast double-click), and the collapsed state must win.
  assert.equal(
    selectExcerptDisplay({ expanded: false, loading: true, excerpt: "some log" }).state,
    EXCERPT_STATE.COLLAPSED,
  );
});

test("loading: expanded with a fetch in flight, before any excerpt has arrived", () => {
  const display = selectExcerptDisplay({ expanded: true, loading: true });
  assert.equal(display.state, EXCERPT_STATE.LOADING);
  assert.deepEqual(display.lines, []);
});

test("error: the lazy GET /v1/jobs/{id} fetch failed", () => {
  const display = selectExcerptDisplay({ expanded: true, error: "not found" });
  assert.equal(display.state, EXCERPT_STATE.ERROR);
  assert.equal(display.message, "not found");
});

test("error takes priority over a stale excerpt from a PREVIOUS successful fetch", () => {
  const display = selectExcerptDisplay({ expanded: true, error: "network", excerpt: "old log" });
  assert.equal(display.state, EXCERPT_STATE.ERROR);
});

test("empty: excerpt is null (job never produced output, e.g. cancelled before it ran)", () => {
  assert.equal(selectExcerptDisplay({ expanded: true, excerpt: null }).state, EXCERPT_STATE.EMPTY);
});

test("empty: excerpt is undefined (fetch never completed, distinct from null but same display)", () => {
  assert.equal(
    selectExcerptDisplay({ expanded: true, excerpt: undefined }).state,
    EXCERPT_STATE.EMPTY,
  );
});

test("empty: excerpt is whitespace-only", () => {
  assert.equal(selectExcerptDisplay({ expanded: true, excerpt: "   \n  " }).state, EXCERPT_STATE.EMPTY);
});

test("ready: a normal multi-line excerpt is split into lines, not truncated", () => {
  const display = selectExcerptDisplay({
    expanded: true,
    excerpt: "line one\nline two\n[vault-api] done",
  });
  assert.equal(display.state, EXCERPT_STATE.READY);
  assert.equal(display.truncated, false);
  assert.deepEqual(display.lines, ["line one", "line two", "[vault-api] done"]);
});

test("ready: a single-line excerpt with no newline at all", () => {
  const display = selectExcerptDisplay({ expanded: true, excerpt: "[vault-api] queued." });
  assert.equal(display.state, EXCERPT_STATE.READY);
  assert.deepEqual(display.lines, ["[vault-api] queued."]);
});

// ---------------------------------------------------------------------
// api/README.md: "log_excerpt is the ANSI-stripped tail... prefixed with
// [...truncated...] when it was cut". The marker must be detected AND
// stripped from the displayed body, and the caller told to show a note —
// the marker text itself must never leak into the first displayed line.
// ---------------------------------------------------------------------
test("ready: a truncated excerpt is flagged AND the marker is stripped from the body", () => {
  const display = selectExcerptDisplay({
    expanded: true,
    excerpt: "[...truncated...]\nthe rest of the log\nmore output",
  });
  assert.equal(display.state, EXCERPT_STATE.READY);
  assert.equal(display.truncated, true);
  assert.deepEqual(display.lines, ["the rest of the log", "more output"]);
  assert.ok(!display.lines.some((l) => l.includes("truncated")));
});

test("ready: the truncation marker is only recognised at the very START of the text", () => {
  // A log line that happens to CONTAIN the words "truncated" elsewhere must
  // not be mistaken for the real marker — only a literal prefix counts.
  const display = selectExcerptDisplay({
    expanded: true,
    excerpt: "line one\n[...truncated...] appears mid-log, not as a real marker",
  });
  assert.equal(display.truncated, false);
  assert.equal(display.lines.length, 2);
});

test("ready: multiple blank lines right after the truncation marker are collapsed away, not shown as empty lines", () => {
  const display = selectExcerptDisplay({ expanded: true, excerpt: "[...truncated...]\n\n\nreal content" });
  assert.equal(display.truncated, true);
  assert.deepEqual(display.lines, ["real content"]);
});

File: app/app/src/test/java/dev/steamvault/app/ui/downloads/logic/LogExcerptTest.kt
package dev.steamvault.app.ui.downloads.logic

import kotlinx.coroutines.test.runTest
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

/** Mirrors `web/tests/log-excerpt.test.js` against `selectExcerptDisplay`'s
 * branch order, plus [ExcerptCache]'s own state-machine (first-expand
 * fetch, cached re-expand, failed fetch allows retry) -- the stateful half
 * `log-excerpt.js` only ever demonstrated inline inside `downloads.js`. */
class LogExcerptTest {

    // -----------------------------------------------------------------
    // selectExcerptDisplay -- pure presentation, no fetch involved.
    // -----------------------------------------------------------------

    @Test
    fun `collapsed when not expanded, regardless of other fields`() {
        val display = selectExcerptDisplay(ExcerptFetchState(expanded = false, excerpt = "some text"))
        assertEquals(ExcerptState.COLLAPSED, display.state)
    }

    @Test
    fun `loading takes priority once expanded`() {
        val display = selectExcerptDisplay(ExcerptFetchState(expanded = true, loading = true))
        assertEquals(ExcerptState.LOADING, display.state)
    }

    @Test
    fun `error state carries the message`() {
        val display = selectExcerptDisplay(ExcerptFetchState(expanded = true, error = "boom"))
        assertEquals(ExcerptState.ERROR, display.state)
        assertEquals("boom", display.message)
    }

    @Test
    fun `empty when the excerpt is blank`() {
        val display = selectExcerptDisplay(ExcerptFetchState(expanded = true, excerpt = ""))
        assertEquals(ExcerptState.EMPTY, display.state)
    }

    @Test
    fun `empty when the excerpt has never been fetched (null, not loading, no error)`() {
        val display = selectExcerptDisplay(ExcerptFetchState(expanded = true))
        assertEquals(ExcerptState.EMPTY, display.state)
    }

    @Test
    fun `ready splits the excerpt into lines`() {
        val display = selectExcerptDisplay(ExcerptFetchState(expanded = true, excerpt = "line one\nline two"))
        assertEquals(ExcerptState.READY, display.state)
        assertEquals(listOf("line one", "line two"), display.lines)
        assertFalse(display.truncated)
    }

    @Test
    fun `ready strips the truncation marker and flags truncated`() {
        val display = selectExcerptDisplay(
            ExcerptFetchState(expanded = true, excerpt = "[...truncated...]\nthe rest\nof it"),
        )
        assertEquals(ExcerptState.READY, display.state)
        assertTrue(display.truncated)
        assertEquals(listOf("the rest", "of it"), display.lines)
    }

    // MUTATION PIN: the truncation check is `startsWith`, anchored to
    // position 0 (vault-api only ever prefixes a cut excerpt with the
    // marker, api/README.md "Queue semantics") -- a mutation to `contains`
    // would still pass every other test in this file, since all of them
    // either have the marker at position 0 or omit it entirely. A real
    // SteamPrefill/vault-api log line can legitimately quote the literal
    // string mid-stream (e.g. echoing a previous run's own truncated
    // excerpt back through some diagnostic path) without the CURRENT fetch
    // actually being cut -- that must NOT be reported as truncated, and the
    // marker text must NOT be stripped out of the middle of the body.
    @Test
    fun `a truncation marker appearing mid-stream, not at position 0, is not treated as truncation`() {
        val body = "first line\n[...truncated...]\nlast line"
        val display = selectExcerptDisplay(ExcerptFetchState(expanded = true, excerpt = body))
        assertEquals(ExcerptState.READY, display.state)
        assertFalse(display.truncated) // MUTATION TARGET: startsWith -> contains would flip this
        assertEquals(listOf("first line", "[...truncated...]", "last line"), display.lines)
    }

    // -----------------------------------------------------------------
    // ExcerptCache -- the lazy-fetch state machine (WP brief: "one JobDetail
    // fetch per job on first expand, cached for the session").
    // -----------------------------------------------------------------

    @Test
    fun `first expand triggers exactly one fetch`() = runTest {
        var fetchCount = 0
        val cache = ExcerptCache { _ ->
            fetchCount++
            Result.success("hello")
        }
        cache.toggleExpanded(1)
        cache.ensureLoaded(1)
        assertEquals(1, fetchCount)
        assertEquals("hello", cache.stateFor(1).excerpt)
    }

    @Test
    fun `a cached re-expand does not refetch (MUTATION TARGET)`() = runTest {
        var fetchCount = 0
        val cache = ExcerptCache { _ ->
            fetchCount++
            Result.success("hello")
        }
        cache.toggleExpanded(1) // expand
        cache.ensureLoaded(1)
        cache.toggleExpanded(1) // collapse
        cache.toggleExpanded(1) // expand again
        cache.ensureLoaded(1)
        assertEquals(1, fetchCount) // still exactly one network call
    }

    @Test
    fun `a genuinely empty log is cached too, not re-fetched`() = runTest {
        var fetchCount = 0
        val cache = ExcerptCache { _ ->
            fetchCount++
            Result.success("")
        }
        cache.toggleExpanded(1)
        cache.ensureLoaded(1)
        cache.ensureLoaded(1) // second call, e.g. a spurious re-invoke
        assertEquals(1, fetchCount)
        assertEquals(ExcerptState.EMPTY, selectExcerptDisplay(cache.stateFor(1)).state)
    }

    @Test
    fun `a failed fetch state is observable and allows a retry on next load`() = runTest {
        var fetchCount = 0
        val cache = ExcerptCache { _ ->
            fetchCount++
            if (fetchCount == 1) Result.failure(RuntimeException("network down")) else Result.success("recovered")
        }
        cache.toggleExpanded(1)
        cache.ensureLoaded(1)
        val afterFailure = cache.stateFor(1)
        assertEquals("network down", afterFailure.error)
        assertEquals(ExcerptState.ERROR, selectExcerptDisplay(afterFailure).state)
        assertFalse(afterFailure.loading)

        // Retry (e.g. collapse/re-expand, or a manual retry action) --
        // excerpt is still null, so the guard must allow a second attempt.
        cache.ensureLoaded(1)
        val afterRetry = cache.stateFor(1)
        assertEquals("recovered", afterRetry.excerpt)
        assertEquals(2, fetchCount)
    }

    @Test
    fun `different jobs have independent cache entries`() = runTest {
        val cache = ExcerptCache { jobId -> Result.success("excerpt for $jobId") }
        cache.toggleExpanded(1)
        cache.ensureLoaded(1)
        cache.toggleExpanded(2)
        cache.ensureLoaded(2)
        assertEquals("excerpt for 1", cache.stateFor(1).excerpt)
        assertEquals("excerpt for 2", cache.stateFor(2).excerpt)
    }

    @Test
    fun `a never-toggled job reports the default collapsed state`() {
        val cache = ExcerptCache { Result.success("unused") }
        val state = cache.stateFor(42)
        assertFalse(state.expanded)
        assertEquals(ExcerptState.COLLAPSED, selectExcerptDisplay(state).state)
    }
}
